File: dotnet/src/Agents/Core/Chat/AggregatorTerminationStrategy.cs
// Copyright (c) Microsoft. All rights reserved.
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.SemanticKernel.Agents.Chat;

/// <summary>
/// Defines aggregation behavior for <see cref="AggregatorTerminationStrategy"/>.
/// </summary>
[Experimental("SKEXP0110")]
public enum AggregateTerminationCondition
{
    /// <summary>
    /// All aggregated strategies must agree on termination.
    /// </summary>
    All,

    /// <summary>
    /// Any single aggregated strategy will terminate.
    /// </summary>
    Any,
}

/// <summary>
/// Provides methods to aggregate a set of <see cref="TerminationStrategy"/> objects.
/// </summary>
/// <param name="strategies">The set of strategies upon which to aggregate.</param>
[Experimental("SKEXP0110")]
public sealed class AggregatorTerminationStrategy(params TerminationStrategy[] strategies) : TerminationStrategy
{
    private readonly TerminationStrategy[] _strategies = strategies;

    /// <summary>
    /// Gets the logical operation for aggregation.
    /// </summary>
    /// <value>
    /// The logical operation for aggregation, which can be <see cref="AggregateTerminationCondition.All"/> or <see cref="AggregateTerminationCondition.Any"/>. The default is <see cref="AggregateTerminationCondition.All"/>.
    /// </value>
    public AggregateTerminationCondition Condition { get; init; } = AggregateTerminationCondition.All;

    /// <inheritdoc/>
    protected override async Task<bool> ShouldAgentTerminateAsync(Agent agent, IReadOnlyList<ChatMessageContent> history, CancellationToken cancellationToken = default)
    {
        this.Logger.LogAggregatorTerminationStrategyEvaluating(nameof(ShouldAgentTerminateAsync), this._strategies.Length, this.Condition);

        var strategyExecution = this._strategies.Select(s => s.ShouldTerminateAsync(agent, history, cancellationToken));

        var results = await Task.WhenAll(strategyExecution).ConfigureAwait(false);
        bool shouldTerminate =
            this.Condition == AggregateTerminationCondition.All ?
                results.All(r => r) :
                results.Any(r => r);

        return shouldTerminate;
    }
}

File: python/semantic_kernel/agents/strategies/termination/aggregator_termination_strategy.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio
from enum import Enum
from typing import TYPE_CHECKING

from pydantic import Field

from semantic_kernel.agents.strategies.termination.termination_strategy import TerminationStrategy
from semantic_kernel.contents.chat_message_content import ChatMessageContent
from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.utils.feature_stage_decorator import experimental

if TYPE_CHECKING:
    from semantic_kernel.agents.agent import Agent


@experimental
class AggregateTerminationCondition(str, Enum):
    """The condition for terminating the aggregation process."""

    ALL = "All"
    ANY = "Any"


@experimental
class AggregatorTerminationStrategy(KernelBaseModel):
    """A strategy that aggregates multiple termination strategies."""

    strategies: list[TerminationStrategy]
    condition: AggregateTerminationCondition = Field(default=AggregateTerminationCondition.ALL)

    async def should_terminate_async(
        self,
        agent: "Agent",
        history: list[ChatMessageContent],
    ) -> bool:
        """Check if the agent should terminate.

        Args:
            agent: The agent to check.
            history: The history of messages in the conversation.

        Returns:
            True if the agent should terminate, False otherwise
        """
        strategy_execution = [strategy.should_terminate(agent, history) for strategy in self.strategies]
        results = await asyncio.gather(*strategy_execution)

        if self.condition == AggregateTerminationCondition.ALL:
            return all(results)
        return any(results)
